File: app/src/main/java/net/simno/dmach/ui/adapter/PatchAdapter.java
/*
* Copyright (C) 2015 Simon Norberg
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

package net.simno.dmach.ui.adapter;

import android.app.Activity;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.simno.dmach.R;
import net.simno.dmach.model.Patch;

import java.util.Collections;
import java.util.List;

import butterknife.BindColor;
import butterknife.ButterKnife;
import io.reactivex.functions.Consumer;

public class PatchAdapter extends RecyclerView.Adapter<PatchViewHolder>
        implements Consumer<List<Patch>>, PatchViewHolder.OnClickListener {

    public interface OnPatchClickListener {
        void onPatchClick(Patch patch);
        void onPatchLongClick(Patch patch);
    }

    @BindColor(R.color.khaki) int khaki;
    @BindColor(R.color.gurkha) int gurkha;

    private final OnPatchClickListener listener;
    private List<Patch> patches = Collections.emptyList();

    public PatchAdapter(Activity activity) {
        this.listener = (OnPatchClickListener) activity;
        ButterKnife.bind(this, activity);
    }

    @Override
    public PatchViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.item_patch, parent, false);
        view.setBackgroundColor(viewType == 1 ? khaki : gurkha);
        return new PatchViewHolder(view, this);
    }

    @Override
    public void onBindViewHolder(PatchViewHolder holder, int position) {
        holder.bindModel(patches.get(position));
    }

    @Override
    public int getItemCount() {
        return patches.size();
    }

    @Override
    public int getItemViewType(int position) {
        return (position % 2 == 0) ? 1 : 0;
    }

    @Override
    public void accept(List<Patch> patches) throws Exception {
        this.patches = patches;
        notifyDataSetChanged();
    }

    @Override
    public void onClick(int position) {
        listener.onPatchClick(patches.get(position));
    }

    @Override
    public void onLongClick(int position) {
        listener.onPatchLongClick(patches.get(position));
    }
}

File: app/src/main/java/net/simno/dmach/patch/PatchAdapter.kt
package net.simno.dmach.patch

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow
import net.simno.dmach.R
import net.simno.dmach.data.Patch
import net.simno.dmach.databinding.PatchItemBinding

class PatchAdapter : PagingDataAdapter<Patch, PatchViewHolder>(DIFF_CALLBACK) {

    private val _clicks = Channel<Patch>()
    private val _longClicks = Channel<Patch>()
    val clicks: Flow<Patch> = _clicks.receiveAsFlow()
    val longClicks: Flow<Patch> = _longClicks.receiveAsFlow()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PatchViewHolder {
        val binding = PatchItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        val background = ContextCompat.getColor(parent.context, viewType)
        return PatchViewHolder(binding, background)
    }

    override fun onBindViewHolder(holder: PatchViewHolder, position: Int) {
        val patch = getItem(position) ?: return
        holder.bind(
            patch = patch,
            onClick = {
                runCatching { _clicks.offer(patch) }
            },
            onLongClick = {
                if (itemCount > 1) {
                    runCatching { _longClicks.offer(patch) }
                }
                true
            }
        )
    }

    override fun getItemViewType(position: Int): Int {
        return when {
            position % 2 == 0 -> R.color.khaki
            else -> R.color.gurkha
        }
    }

    companion object {
        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Patch>() {
            override fun areItemsTheSame(oldItem: Patch, newItem: Patch) = oldItem.title == newItem.title

            override fun areContentsTheSame(oldItem: Patch, newItem: Patch) = oldItem == newItem
        }
    }
}
